lib/src/Pos/Logic/Types.hs

```diff
@@ -0,0 +1,203 @@
+    , recoveryInProgress :: m Bool
+
+    , securityParams     :: SecurityParams
+    }
+
+-- | First iteration solution to the inv/req/data/mempool system.
+-- Diffusion layer will set up the relays, but it needs help from the logic
+-- layer in order to figure out what to request after an inv, what to relay
+-- (the Bool return values), and of course what to actually do with the data,
+-- which is tied up with whether it should be relayed (processing the data
+-- and returning a Bool to indicate whether to relay).
+--
+--   - Does 'toKey' need to be monadic? Surely the datum itself contains the
+--     key and this can be a pure function 'val -> key'. Maybe not? Must look
+--     into it.
+--   - How much help from the logic layer does the diffusion layer really need
+--     in order to determine whether to relay something?
+--     How much of the inv/req/data system can be done entirely by the
+--     diffusion layer? Perhaps it should keep a little cache of keys it has
+--     seen, ignore keys which are in that cache, and for keys which are not,
+--     enqueue them up to the logic layer.
+--     For handling 'Req' messages, though, logic layer support will in fact
+--     be necessary I think; can't expect the diffusion layer to take care of
+--     storing everything.
+--
+--   - Must investigate the mempool aspect.
+--     Only transactions use this.
+--     The MempoolParams KeyMempool constructor includes an m [key] which I
+--     guess produces all keys in the mempool. Ah yes, this is so that a peer
+--     can request everything in the mempool.
+--     See 'handleMempoolL'. The entire mempool (as keys) is dumped to a list,
+--     and then each key is 'Inv'd once (InvMsg sent).
+--
+--     I do not believe we ever make a mempool request (MempoolMsg).
+--     Ok we can probably dump this.
+data KeyVal key val m = KeyVal
+    { toKey :: val -> m key
+    , handleInv :: key -> m Bool
+    , handleReq :: key -> m (Maybe val)
+    , handleData :: val -> m Bool
+    }
+
+-- | Failure description for getting a block from the logic layer.
+data GetBlockError = GetBlockError Text
+
+deriving instance Show GetBlockError
+instance Exception GetBlockError
+
+-- | Failure description for getting a block header from the logic layer.
+data GetBlockHeaderError = GetBlockHeaderError Text
+
+deriving instance Show GetBlockHeaderError
+instance Exception GetBlockHeaderError
+
+-- | Failure description for getting a block header from the logic layer.
+data GetBlockHeadersError = GetBlockHeadersError Text
+
+deriving instance Show GetBlockHeadersError
+instance Exception GetBlockHeadersError
+
+-- | Failure description for getting the tip of chain from the logic layer.
+data GetTipError = GetTipError Text
+
+deriving instance Show GetTipError
+instance Exception GetTipError
+
+-- | A diffusion layer: its interface, and a way to run it.
+data LogicLayer m = LogicLayer
+    { runLogicLayer :: forall x . m x -> m x
```

Comment:
`m` in negative position again :disappointed:

Reply:
it's really hard to reason about manual dictionaries with `m` in negative positions when it comes to unlifting (I go into more details in CSL-1823)

Reply:
So long as we never pick a concrete `m`, we don't have to do lifts or natural transformations.

Notice that the full diffusion layer implementation leaves `m` free subject only to some constraints `DiffusionWorkMode m, MonadFix m`. We can pick the concrete `m` and let GHC solve the constraints; no need to apply natural transformations.

You don't change the implementation (logic or diffusion layer record) in order to accommodate your choice of monad; you change your choice of monad to accommodate the implementation.